windows-agent/internal/daemon/daemon_test.go

```diff
@@ -144,35 +145,73 @@ func TestStartQuit(t *testing.T) {
-	// affects the behavior of the getWslIP function.
-	daemon.SetWslIPErr(t)
-
-	addrDir := t.TempDir()
+func TestServeWSLIP(t *testing.T) {
+	t.Parallel()
 
 	registerer := func(context.Context) *grpc.Server {
 		return grpc.NewServer()
 	}
 
-	// Very lenient timeout because we expect the Serve to fail immediately.
-	// If it doesn't, the test will fail due to the context timeout (otherwise
-	// it would hang indefinitely).
-	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
-	defer cancel()
+	testcases := map[string]struct {
+		netmode             string
+		denyWslSystemAccess bool
+		withAdapters        daemon.MockIPAdaptersState
 
-	d := daemon.New(ctx, registerer, addrDir)
-	defer d.Quit(ctx, false)
+		wantErr bool
+	}{
+		"Success":                       {withAdapters: daemon.Ok},
+		"With a single Hyper-V Adapter": {withAdapters: daemon.SingleHyperVAdapterInList},
+		"With mirrored networking mode": {netmode: "mirrored", withAdapters: daemon.Ok},
+		"With no access to wsl --system but net mode is the default (NAT)": {denyWslSystemAccess: true, withAdapters: daemon.Ok},
+
+		"Error when the networking mode is unknown":       {netmode: "unknown", wantErr: true},
+		"Error when the list of adapters is empty":        {withAdapters: daemon.EmptyList, wantErr: true},
+		"Error when there is no Hyper-V adapter the list": {withAdapters: daemon.NoHyperVAdapterInList, wantErr: true},
+	}
 
-	err := d.Serve(ctx)
-	require.Error(t, err, "Serve should fail when the WSL IP cannot be found")
-
-	select {
-	case <-ctx.Done():
-		// Most likely, Serve did not fail and instead started serving,
-		// only to be stopped by the test timeout.
-		require.Fail(t, "Serve should have failed immediately")
-	default:
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			addrDir := t.TempDir()
+			// Very lenient timeout because we either expect Serve to fail immediately or we stop it manually.
+			// As the last resource, the test will fail due to the context timeout (otherwise it would hang indefinitely).
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+			defer cancel()
+
+			d := daemon.New(ctx, registerer, addrDir)
+
+			wsl := daemon.NewWslSystemMock(tc.netmode, nil, tc.denyWslSystemAccess)
+			ipconfig := daemon.NewHostIPConfigMock(tc.withAdapters)
+			serveErr := make(chan error)
+			go func() {
+				serveErr <- d.Serve(ctx, daemon.WithWslSystemDistro(wsl), daemon.WithHostIPConfig(&ipconfig))
+			}()
+
+			if tc.wantErr {
+				require.NotErrorIs(t, <-serveErr, grpc.ErrServerStopped, "Serve should fail when the WSL IP cannot be found")
+				return
+			}
+
+			serverStopped := make(chan struct{})
+			go func() {
+				time.Sleep(500 * time.Millisecond)
+				d.Quit(ctx, false)
+				close(serverStopped)
+			}()
+			<-serverStopped
+
+			require.ErrorIs(t, <-serveErr, grpc.ErrServerStopped, "Serve should not fail when the WSL IP is found")
```

Comment:
I see that you rely on the exact error for stopping now that you wrap it. Does it bring anything in the rest of the code? It seems that apart from the test, we don’t rely on this exact error (and so, don’t really care as part of our API).

Shouldn’t we just want some error being returned in both case, when the WSL IP cannot be found? And maybe in the `Serve` method, we capture the underlying error and tell "if we get ErrServerStopped", this is ok? (honest question, needs digging)

Reply:
API wise we don't need to return that specific error object. And, with the latest changes, we either fail early (thus with a different error) or shouldn't fail, for which case `require.NoError` must do it.